File: gateway/src/main/java/com/beat/gateway/refreshtoken/internal/RefreshTokenService.java
package com.beat.gateway.refreshtoken.internal;

import java.util.OptionalLong;

import com.beat.contracts.auth.RefreshTokenPort;
import com.beat.gateway.refreshtoken.internal.store.RefreshToken;
import com.beat.gateway.refreshtoken.internal.store.RefreshTokenRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

@Slf4j
@RequiredArgsConstructor
@Service
public class RefreshTokenService implements RefreshTokenPort {

	private final RefreshTokenRepository refreshTokenRepository;

	@Override
	public void saveRefreshToken(final long memberId, final String refreshToken) {
		refreshTokenRepository.save(RefreshToken.of(memberId, refreshToken));
	}

	@Override
	public OptionalLong findMemberIdByRefreshToken(final String refreshToken) {
		return refreshTokenRepository.findByRefreshToken(refreshToken)
			.map(token -> OptionalLong.of(token.getId()))
			.orElseGet(OptionalLong::empty);
	}

	@Override
	public boolean deleteRefreshToken(final long memberId) {
		return refreshTokenRepository.findById(memberId).map(token -> {
			refreshTokenRepository.delete(token);
			log.info("Deleted refresh token for memberId={}", memberId);
			return true;
		}).orElse(false);
	}
}

File: gateway/src/main/kotlin/com/beat/gateway/refreshtoken/internal/RefreshTokenService.kt
package com.beat.gateway.refreshtoken.internal

import com.beat.contracts.auth.RefreshTokenPort
import com.beat.gateway.refreshtoken.internal.store.RefreshToken
import com.beat.gateway.refreshtoken.internal.store.RefreshTokenRepository
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.util.OptionalLong

@Service
class RefreshTokenService(
    private val refreshTokenRepository: RefreshTokenRepository,
) : RefreshTokenPort {

    private val log = LoggerFactory.getLogger(RefreshTokenService::class.java)

    override fun saveRefreshToken(memberId: Long, refreshToken: String) {
        refreshTokenRepository.save(RefreshToken(memberId, refreshToken))
    }

    override fun findMemberIdByRefreshToken(refreshToken: String): OptionalLong =
        refreshTokenRepository.findByRefreshToken(refreshToken)
            .map { token -> OptionalLong.of(token.id) }
            .orElseGet { OptionalLong.empty() }

    override fun deleteRefreshToken(memberId: Long): Boolean =
        refreshTokenRepository.findById(memberId).map { token ->
            refreshTokenRepository.delete(token)
            log.info("Deleted refresh token for memberId={}", memberId)
            true
        }.orElse(false)
}
